job/external_dependency_resolver.go

```diff
@@ -0,0 +1,195 @@
+package job
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/odpf/optimus/config"
+	"github.com/odpf/optimus/ext/resourcemgr"
+	"github.com/odpf/optimus/models"
+	"github.com/odpf/optimus/store"
+)
+
+type ExternalDependencyResolver interface {
+	FetchInferredExternalDependenciesPerJobName(context.Context, models.ProjectID) (map[string]models.ExternalDependency, error)
+	FetchStaticExternalDependenciesPerJobName(context.Context, models.ProjectID) (map[string]models.ExternalDependency, []models.UnknownDependency, error)
+}
+
+type externalDependencyResolver struct {
+	optimusResourceManagers        []resourcemgr.ResourceManager
+	unknownJobDependencyRepository store.UnknownJobDependencyRepository
+}
+
+// NewExternalDependencyResolver creates a new instance of externalDependencyResolver
+func NewExternalDependencyResolver(resourceManagerConfigs []config.ResourceManager, unknownJobDependencyRepository store.UnknownJobDependencyRepository) (ExternalDependencyResolver, error) {
+	var optimusResourceManagers []resourcemgr.ResourceManager
+	for _, conf := range resourceManagerConfigs {
+		switch conf.Type {
+		case "optimus":
+			getter, err := resourcemgr.NewOptimusResourceManager(conf)
+			if err != nil {
+				return nil, err
+			}
+			optimusResourceManagers = append(optimusResourceManagers, getter)
+		default:
+			return nil, fmt.Errorf("resource manager [%s] is not recognized", conf.Type)
+		}
+	}
+	return &externalDependencyResolver{
+		unknownJobDependencyRepository: unknownJobDependencyRepository,
+		optimusResourceManagers:        optimusResourceManagers,
+	}, nil
+}
+
+func (e *externalDependencyResolver) FetchInferredExternalDependenciesPerJobName(ctx context.Context, projectID models.ProjectID) (map[string]models.ExternalDependency, error) {
+	if ctx == nil {
+		return nil, errors.New("context is nil")
+	}
+	unknownInferredDependenciesPerJobName, err := e.unknownJobDependencyRepository.GetUnknownInferredDependencyURNsPerJobName(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+	jobSpecFiltersPerJobName := e.toJobSpecFiltersPerJobNameForResourceNames(unknownInferredDependenciesPerJobName)
+	return e.fetchInferredExternalDependencyPerJobName(ctx, jobSpecFiltersPerJobName)
+}
+
+func (e *externalDependencyResolver) FetchStaticExternalDependenciesPerJobName(ctx context.Context, projectID models.ProjectID) (map[string]models.ExternalDependency, []models.UnknownDependency, error) {
+	if ctx == nil {
+		return nil, nil, errors.New("context is nil")
+	}
+	unknownStaticDependenciesPerJobName, err := e.unknownJobDependencyRepository.GetUnknownStaticDependencyNamesPerJobName(ctx, projectID)
+	if err != nil {
+		return nil, nil, err
+	}
+	jobSpecFiltersPerJobName, err := e.toJobSpecFiltersPerJobNameForDependencyNames(unknownStaticDependenciesPerJobName)
```

Comment:
function names are hard to understand especially in the dependency resolution

Reply:
this will be removed as part of previous feedback.

Reply:
removed